### PR TITLE
[build.ps1/setup.sh] do not export vcpkg_root anymore, add setup scripts to CI

### PR DESCRIPTION
**Pull request by cenit**
## [build.ps1] improvements

in order to have a clean shell after build.
It might help with very peculiar circumstances when for example you clone darknet, build it enabling vcpkg but without having it already on the system, then delete the whole darknet folder and then re-try building it again enabling vcpkg, all without ever closing and reopening the shell.

I know that it's a very corner case which should never happen, but there is a user that tried it and it didn't work. I hope it can be fixed with this PR

## [setup.sh/ps1] added in CI testing

in order to verify and monitor that they are working ok
We might even switch to them to provision CI machines

### CODE REVIEW

**Comment by AlexeyAB** [comment]
@cenit Hi,

Also can we somehow remove the need for this command? Or add it to the build.dps1 script? `Set-ExecutionPolicy unrestricted`  https://github.com/AlexeyAB/darknet/pull/7644#issuecomment-839145728 
So that people who are not familiar with Powershell do not have to search the Internet for a way to solve the problem: 
`+ CategoryInfo :Security error: (:) [], PSSecurityException + FullyQualifiedErrorId : UnauthorizedAccess`

**Comment by cenit** [comment]
unfortunately not.
You cannot unblock this security mechanism from within the script itself.
The default on macOS and on Linux is unrestricted, but unfortunately on Windows they choose restricted, so before running scripts you have either to unrestrict the shell or at least to unblock that single script with
Unblock-File -Path .\build.ps1

We can add a line in readme to highlight the command for people still having a restricted powershell

File: README.md [hunk 1]
```diff
@@ -355,50 +355,67 @@ If you open the `build.ps1` script at the beginning you will find all available
...
 ### How to compile on Windows (using `vcpkg`)
 
 This is the recommended approach to build Darknet on Windows.
 
-1. Install Visual Studio 2017 or 2019. In case you need to download it, please go here: [Visual Studio Community](http://visualstudio.com). Remember to install English language pack, this is mandatory for vcpkg!
+1. Open Powershell as Administrator (Start -> All programs -> right click on Windows Powershell and choose "Run as administrator") and type this command:
 
-2. Install CUDA (at least v10.0) enabling VS Integration during installation.
+```PowerShell
+PS />              Set-ExecutionPolicy unrestricted
```
Comment (cenit) [inline README.md:399]: @AlexeyAB  this line should help users

Comment (cenit) [reply README.md:399]: sorry for the noise in the README. The linter cleaned up a lot of mismatches and I didn't want to stop it 🤣

**Comment by cenit** [comment]
> Can we avoid it by runing `build.ps1` as Administrator (right mouse click -> run as Administrator) ?
> Or can we create `build.cmd` script with content `Unblock-File -Path .\build.ps1` ? Should this file `build.cmd` be executed as Administrator?

from cmd you cannot run a powershell command (unblock-file or set-executionpolicy)

no need to run anything as an administrator

The easiest method (now included in the README in this PR) is to write

Set-ExecutionPolicy unrestricted -Scope CurrentUser

in a normal Windows PowerShell to unblock it

File: README.md [hunk 2]
```diff
@@ -355,50 +355,67 @@ If you open the `build.ps1` script at the beginning you will find all available
...
 
 This is the recommended approach to build Darknet on Windows.
 
-1. Install Visual Studio 2017 or 2019. In case you need to download it, please go here: [Visual Studio Community](http://visualstudio.com). Remember to install English language pack, this is mandatory for vcpkg!
+1. Open Powershell (Start -> All programs -> right click on Windows Powershell) and type this command:
 
-2. Install CUDA (at least v10.0) enabling VS Integration during installation.
+```PowerShell
+PS />               Set-ExecutionPolicy unrestricted -Scope CurrentUser
+```
```
Comment (cenit) [inline README.md:400]: here the added comments about running the "unblock" snippet first of all

**Comment by cenit** [comment]
tomorrow i will add a nicer error message in case the execution policy is not set correctly, so that unexperienced users are better guided to fix it

**Comment by cenit** [comment]
@AlexeyAB error is due to circleci internal problems. PR is ready for me

**Comment by AlexeyAB** [comment]
@cenit Thanks!

Also I think we should change
* Add `cmake . && make -j8` for Cmake installation.
* Remove Powershell from Cmake installation.


```
How to compile on Linux/macOS (using CMake)
The CMakeLists.txt will attempt to find installed optional dependencies like CUDA, cudnn, ZED and build against those. It will also create a shared object library file to use darknet for code development.

Install powershell if you do not already have it (guide here).

To update CMake on Ubuntu, it's better to follow guide here: https://apt.kitware.com/
```

to this
```
How to compile on Linux/macOS (using CMake)

cmake .
make -j8

The CMakeLists.txt will attempt to find installed optional dependencies like CUDA, cudnn, ZED and build against those. It will also create a shared object library file to use darknet for code development.


To update CMake on Ubuntu, it's better to follow guide here: https://apt.kitware.com/
```

**Comment by stephanecharette** [comment]
Installing Microsoft Powershell on linux devices is definitely a non-starter for many us like myself who work in a pure linux environment.  There are objections to this at my office, and I personally object as well.  Using `make` and/or `cmake` on linux should not require a Microsoft shell.

**Comment by cenit** [comment]
> Installing Microsoft Powershell on linux devices is definitely a non-starter for many us like myself who work in a pure linux environment.  There are objections to this at my office, and I personally object as well.  Using `make` and/or `cmake` on linux should not require a Microsoft shell.

it does not become impure 🤣 
in any case it was just for inexperienced users, also on windows it's not required if you know what you are doing. License is MIT and so does not put any obstacle

but ok i will put different instructions per different OSes. I like being able to run same commands everywhere, but tastes are personal

**Comment by cenit** [comment]
> @cenit Thanks!
> 
> Also I think we should change
> 
> * Add `cmake . && make -j8` for Cmake installation.
> * Remove Powershell from Cmake installation.
> 
> ```
> How to compile on Linux/macOS (using CMake)
> The CMakeLists.txt will attempt to find installed optional dependencies like CUDA, cudnn, ZED and build against those. It will also create a shared object library file to use darknet for code development.
> 
> Install powershell if you do not already have it (guide here).
> 
> To update CMake on Ubuntu, it's better to follow guide here: https://apt.kitware.com/
> ```
> 
> to this
> 
> ```
> How to compile on Linux/macOS (using CMake)
> 
> cmake .
> make -j8
> 
> The CMakeLists.txt will attempt to find installed optional dependencies like CUDA, cudnn, ZED and build against those. It will also create a shared object library file to use darknet for code development.
> 
> 
> To update CMake on Ubuntu, it's better to follow guide here: https://apt.kitware.com/
> ```

ok, done in https://github.com/AlexeyAB/darknet/pull/7751/commits/214bac74096d68ccca4dec2875d111a3267c3560

**Comment by AlexeyAB** [comment]
@cenit Thanks!

* I added minor fix: https://github.com/AlexeyAB/darknet/pull/7751/commits/cbe2015724de7868ba82d4fd9839f48ba86b8e01

* Do I understand correctly, that build.ps1/VCPKG can install OpenCV automatically, but can't install CUDA/cuDNN automatically?

* Is it
```
git clone https://github.com/AlexeyAB/darknet
cd darknet
mkdir build_release
cd build_release
cmake ..
cmake --build . --parallel 8
```

better than?
```
git clone https://github.com/AlexeyAB/darknet
cd darknet
cmake .
make -j8
```


Will it create `./darknet` executable file in the `/darknet/build_release` or `/darknet` folder?
Will it copy folders `/cfg/`, `/data/` to the `/build_release` folder automatically?

**Comment by cenit** [comment]
> Do I understand correctly, that build.ps1/VCPKG can install OpenCV automatically, but can't install CUDA/cuDNN automatically?

yes, exactly. That is possible with script/setup.sh and script/setup.ps1, but those scripts requires sudo privileges

> better than?

yes. You should always build out of tree, at least in this way we are building in a subfolder.

I forgot to write (added in https://github.com/AlexeyAB/darknet/pull/7751/commits/4025af1976354837f1c7b9ebafa1ad7889112f3f) the `--target install` command. The target install will put darknet executable and libraries in the darknet main folder in the end, so that the user can also completely delete the build_release folder without loosing anything.

**Comment by AlexeyAB** [comment]
Great!

> > Do I understand correctly, that build.ps1/VCPKG can install OpenCV automatically, but can't install CUDA/cuDNN automatically?
> 
> yes, exactly. That is possible with script/setup.sh and script/setup.ps1, but those scripts require sudo privileges

Can we just type:
`sudo script/setup.sh` and it will automatically install on Linux: `Cmake, PowerShell, VCPKG, CUDA, cuDNN, OpenCV, Darknet` without any question?

**Comment by cenit** [comment]
setup scripts should require some more testing. At the moment they are not verified in CI and so they might not work anymore

we can focus on them in the future if you want. I will start adding some tests on them, but maybe it's better in another PR?

**Comment by cenit** [comment]
anyway, I just looked at setup scripts. They might be still working and they do not require sudo. They will ask for sudo privileges while running, though. This is to avoid running all operations with sudo privileges, which might create problems afterwards (people unable to delete files without sudo, etc)

**Comment by cenit** [comment]
I will add a couple of CI tests for them to be verified and some instructions in the README. Let me know if you want them in this PR or in a separate one

**Comment by cenit** [comment]
@AlexeyAB yes, they should work that way (and are tested in CI now).
Please let me know if you have any other question

**Comment by cenit** [comment]
@AlexeyAB yes, they should work that way (and are tested in CI now). 
You should also avoid running them as sudo, they will ask for sudo permission for specific operations when required, in order to avoid operations to be done with sudo privileges which might cause problem later on (users unable to delete files without sudo permissions, for example)

In Linux the scripts/setup.sh installs everything, on Windows scripts/setup.ps1 installs everything but CUDNN is left out, because we cannot download it without user registration on the website.
Please let me know if you have any other question

**Comment by cenit** [comment]
They are not described in the readme as of now, because we said to keep them a little bit behind the scenes in order to gather some feedback. Is there anyone using them?

**Comment by cenit** [comment]
@AlexeyAB if CI is green for me PR is ready to merge

**Comment by cenit** [comment]
it's green :)

**Comment by AlexeyAB** [comment]
@cenit Thanks, great!

**Pull request merged**